exercises/practice/react/.meta/src/reference/java/React.java

```diff
@@ -0,0 +1,124 @@
+import java.util.*;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+public class React {
+
+    interface Cell<T> {
+        T getValue();
+        default T getTentativeValue() {
+            return getValue();
+        }
+        void addDependent(ComputeCell<T> dependent);
+    }
+
+    interface InputCell<T> extends Cell<T> {
+        void setValue(T newValue);
+    }
+
+    interface ComputeCell<T> extends Cell<T> {
+        void prepareTentativeUpdate();
+        void commitTentativeUpdate();
+        void addCallback(Consumer<T> callback);
+        void removeCallback(Consumer<T> callback);
+    }
+
+    static class CellImpl<T> implements Cell<T> {
```

Comment:
I'm wondering if there is a real need for this 'base' class. It seems to mix the concepts of the input and compute cells, and adds functionality to both subclasses that is only intended for a single subclass:

- The `ComputeCellImpl` class has a method `setValue` even though this should not be possible for a computed cell
- The `InputCellImpl` class has methods to add/remove callbacks and dependent cells, which doesn't make a lot of sense.

The only 'real' shared functionality of both types is already in the original interface: a getter for the cell value. IMO that's not even enough to warrant a base class at all.

Reply:
There is more than `getValue()` that is shared here - there is also adding and tracking `ComputeCells` that depend on it.